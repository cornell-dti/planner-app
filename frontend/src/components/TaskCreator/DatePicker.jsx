// @flow strict

import React from 'react';
import { Calendar } from 'react-calendar';
import { Icon } from 'semantic-ui-react';
import styles from './Picker.css';
import { date2String } from '../../util/datetime-util';
import { NONE_TAG } from '../../util/tag-util';

type Props = {|
  +onDateChange: (date: Date) => void;
  +date: Date;
  +opened: boolean;
  +datePicked: boolean;
  +onPickerOpened: () => void;
|};

export default function DatePicker(props: Props) {
  const {
    date, opened, datePicked, onDateChange, onPickerOpened,
  } = props;
  // Controllers
  const clickPicker = () => { onPickerOpened(); };
  const reset = (e) => { e.stopPropagation(); onDateChange(null); };
  // Nodes
  const displayedNode = (isDefault: boolean) => {
    const style = isDefault ? {} : { background: NONE_TAG.color };
    const internal = isDefault
      ? (<Icon name="calendar outline" className={styles.CenterIcon} />)
      : (
        <React.Fragment>
          <span className={styles.DateDisplay}>{date2String(date)}</span>
          <button type="button" className={styles.ResetButton} onClick={reset}>&times;</button>
        </React.Fragment>
      );
    return (
      <span role="presentation" onClick={clickPicker} className={styles.Label} style={style}>
        {internal}
      </span>
    );
  };
  return (
    <div className={styles.Main}>
      {displayedNode(!datePicked)}
      {opened && (
        <div className={styles.NewTaskDatePick}>
          <Calendar
            onChange={onDateChange}
            value={date}
            minDate={new Date()}
          />
        </div>
      )}
    </div>
  );
}
