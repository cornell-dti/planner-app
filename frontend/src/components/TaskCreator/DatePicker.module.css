
.SelectTypeWrap{
  text-align: right;
  margin:0;
}

.SelectType{
  padding: 2px 5px;
  margin-bottom:15px;
  border-radius:5px;
  cursor:pointer;
  border:2px #5a5a5a solid;
  color:#5a5a5a;
}


.SubtleBtn{
  background:#eee;
  color:#5a5a5a;
  border:none;
  cursor:pointer;
  border-radius:5px;
}
