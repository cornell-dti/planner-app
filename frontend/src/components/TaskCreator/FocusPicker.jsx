// @flow strict

import React from 'react';
import Pin from '../../assets/svgs/pin-2-dark-outline.svg';
import styles from './Picker.css';

type Props = {|
  +pinned: boolean;
  +onPinChange: (inFocus: boolean) => void;
|};

export default function FocusPicker({ pinned, onPinChange }: Props) {
  const clickPicker = (e: SyntheticEvent<HTMLElement>) => {
    e.stopPropagation();
    onPinChange(!pinned);
  };
  const iconStyle = { /* color: pinned ? 'black' : 'gray' */
    backgroundImage: `url(${pinned ? Pin : Pin})`,
  };
  return (
    <div className={styles.Main}>
      <span role="presentation" onClick={clickPicker} className={styles.Label}>
        <Pin className={styles.CenterIcon} />
      </span>
    </div>
  );
}
