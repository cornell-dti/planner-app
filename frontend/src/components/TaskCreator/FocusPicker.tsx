import React, { ReactElement, SyntheticEvent, KeyboardEvent } from 'react';
import clsx from 'clsx';
import styles from './Picker.module.scss';
import SamwiseIcon from '../UI/SamwiseIcon';

type Props = {
  readonly pinned: boolean;
  readonly onPinChange: (inFocus: boolean) => void;
};

export default function FocusPicker({ pinned, onPinChange }: Props): ReactElement {
  const clickPicker = (e: SyntheticEvent<HTMLElement>): void => {
    e.stopPropagation();
    onPinChange(!pinned);
  };
  const pressedPicker = (e: KeyboardEvent): void => {
    if (e.key === 'Enter' || e.key === ' ') {
      e.stopPropagation();
      onPinChange(!pinned);
    }
  };
  const pinIconName = pinned ? 'pin-dark-filled' : 'pin-dark-outline';
  return (
    <div className={styles.Main}>
      <button
        type="button"
        onClick={clickPicker}
        onKeyPress={pressedPicker}
        className={clsx(styles.TestBorder, styles.Label)}
      >
        <SamwiseIcon iconName={pinIconName} className={clsx(styles.CenterIcon, styles.ImageSize)} />
      </button>
    </div>
  );
}
