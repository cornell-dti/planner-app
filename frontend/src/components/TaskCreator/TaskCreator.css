.NewTaskComponent {
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1px solid gray;

  width: 100%;
  padding: 10px 20px;
  text-align: left;

  font-size: 20px;
  font-family: 'Open Sans', sans-serif;
  transition: border 0.3s ease;
}

.NewTaskComponent::placeholder {
  font-weight: lighter;
  color: gray;
  text-align: center;
}

.NewTaskComponent:focus::placeholder {
  color: white;
}

.NewTaskWrap {
  position: relative;
  left: 50%;
  text-align: center;
  transform: translateX(-50%);
  z-index: 5;

  height: initial;
  max-height: 44px;
  margin-bottom: 20px;
}

@media only screen and (min-width: 800px) {
  .NewTaskWrap {
    width: 30vw;
  }
}

@media only screen and (max-width: 800px) {
  .NewTaskWrap {
    width: 90vw;
  }
}


.NewTaskComponent:focus {
  border-bottom: 2px solid black;
  outline: none;
}


.CloseNewTask {
  position: fixed;
  background: black;
  opacity: 0.4;
  z-index: 4;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
}

/**
 * Wrapper for all options only visible when adding task
 */
.NewTaskActive {
  position: absolute;
  top: 12px;
  height: 20px;
  width: 100%;
  right: 0;
  padding-right: 40px;
  text-align: right;
  pointer-events: none;
}

.NewTaskActive * {
  pointer-events: auto;
  vertical-align:top;
}

.OpenNewTask .NewTaskActive {
  display: block;
}


.SubmitNewTask {
  position: absolute;
  right: 0;
  top: -3px;
  width: 20px;
  height: 100%;
  background: white;
  border: none;
  cursor: pointer;
  padding: 0;
  margin: 0;
  font-size: 25px;
  z-index: 0;
}


/*
 * Pop Open Modal
 */
.NewTaskModal {
  background: white;
  position: absolute;
  top: -30px;
  right: -30px;
  padding: 70px 0 10px 0;
  z-index: -1;
  /*box-shadow: 0 2px 25px 5px #444;*/
  box-shadow: 0 8px 65px 3px #444;
  text-align: center;
}
@media only screen and (min-width: 800px) {
  .NewTaskModal {
    width: calc(30vw + 60px);
  }
}

@media only screen and (max-width: 800px) {
  .NewTaskModal {
    width: calc(90vw + 60px);
  }
}

.NewTaskModal ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.NewTaskModal input[type="text"] {
  border-top: none;
  border-left: none;
  border-right: none;
  border-bottom: 1px gray solid;

  width: calc(100% - 60px);
  padding: 10px 20px 10px 50px;

  outline: none;
}

.NewTaskModal input[type="text"]::placeholder {
  color: gray;
}

.NewTaskModal .AddSubTask::before {
  content: '+';
  position: relative;
  font-size: 20px;
}

.NewTaskModal li {
  position: relative;
}


.NewTaskModal li button {
  position: absolute;
  width: 20px;
  height: 20px;
  bottom: 10px;
  border: none;
  background: none;
  left: 40px;
  cursor: pointer;
}

.NewTaskModal i {
  position: absolute;
  font-size: 30px;
  color: gray;
  bottom: 8px;
}

.NewTaskModal li i {
  display: block;
  position: absolute;
  width: 100%;
  height: 100%;
  bottom: 0;
  left: 0;
  margin: 0;
  padding: 0;
  font-size: 15px;
}


.ResetButton {
  position: absolute;
  background: none;
  border: none;
  cursor: pointer;
  bottom: -30px;
  right: 10px;
  color: white;
}

.ResetButton:hover {
  background: rgba(0, 0, 0, 0.2);
}

.ResetButton:active {
  background: none;
}
