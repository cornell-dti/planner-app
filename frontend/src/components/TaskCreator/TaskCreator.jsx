// @flow strict

import React from 'react';
import type { Node } from 'react';
import { Icon } from 'semantic-ui-react';
import 'react-toastify/dist/ReactToastify.css';
import styles from './TaskCreator.css';
import TagPicker from './TagPicker';
import DatePicker from './DatePicker';
import FocusPicker from './FocusPicker';
import { randomId } from '../../util/general-util';
import { dispatchConnect } from '../../store/react-redux-util';
import type { Task, SubTask } from '../../store/store-types';
import {
  addTask as addTaskAction,
  removeTask as removeTaskAction,
} from '../../store/actions';
import type { AddNewTaskAction, RemoveTaskAction } from '../../store/action-types';
import { NONE_TAG_ID } from '../../util/tag-util';
import { replaceSubTask } from '../../util/task-util';
import { isToday } from '../../util/datetime-util';

type Props = {|
  // subscribed from dispatcher.
  +addTask: (task: Task) => AddNewTaskAction;
  +removeTask: (taskId: number) => RemoveTaskAction;
|};

type State = {|
  ...Task;
  +opened: boolean;
  +tagPickerOpened: boolean;
  +datePickerOpened: boolean;
  +datePicked: boolean;
  +needToSwitchFocus: boolean;
|};

/**
 * The placeholder text in the main task input box.
 * @type {string}
 */
const PLACEHOLDER_TEXT = 'What do you have to do?';
/**
 * Generate the initial state.
 * @return {State}
 */
const initialState = (): State => ({
  id: randomId(),
  name: '',
  tag: NONE_TAG_ID, // the id of the None tag.
  date: new Date(),
  complete: false,
  inFocus: false,
  subtasks: [],
  opened: false,
  tagPickerOpened: false,
  datePickerOpened: false,
  datePicked: false,
  needToSwitchFocus: false,
});

class TaskCreator extends React.PureComponent<Props, State> {
  state: State = initialState();

  /*
   * --------------------------------------------------------------------------------
   * Part 1: Openers & Closers
   * --------------------------------------------------------------------------------
   */

  /**
   * Open the new task editor.
   */
  openNewTask = () => this.setState({ opened: true });

  /**
   * Close (collapse) the new task editor.
   */
  closeNewTask = () => this.setState({ opened: false });

  /**
   * Open the tag picker and close the date picker.
   */
  openTagPicker = () => this.setState({ tagPickerOpened: true, datePickerOpened: false });

  /**
   * Open the date picker and close the tag picker.
   */
  openDatePicker = () => this.setState({ tagPickerOpened: false, datePickerOpened: true });

  /*
   * --------------------------------------------------------------------------------
   * Part 2: Manager functions when finished editing.
   * --------------------------------------------------------------------------------
   */

  /**
   * Focus on the task name, if possible.
   */
  focusTaskName = () => {
    if (this.addTask) {
      this.addTask.focus();
    }
  };

  /**
   * Handle on potential save.
   *
   * @param e the event that signals a potential save action.
   */
  handleSave = (e?: SyntheticEvent<HTMLElement>) => {
    if (e != null) {
      e.preventDefault();
    }
    const {
      id, name, tag, date, complete, inFocus, subtasks,
    } = this.state;
    const { addTask } = this.props;
    if (name === '') {
      return;
    }
    const newSubTasks = subtasks.filter(subTask => subTask.name !== '');
    const autoInFocus = inFocus || isToday(date); // Put task in focus is the due date is today.
    const newTask = {
      id, name, tag, date, complete, inFocus: autoInFocus, subtasks: newSubTasks,
    };
    // Add the task to the store.
    addTask(newTask);
    // Reset the state.
    this.setState({ ...initialState() });
    this.closeNewTask();
    if (this.addTask) {
      this.addTask.blur();
    }
  };

  /*
   * --------------------------------------------------------------------------------
   * Part 3: Various Editors
   * --------------------------------------------------------------------------------
   */

  /**
   * Edit the task name.
   *
   * @param e the event that contains the new task name.
   */
  editTaskName = (e: SyntheticEvent<HTMLInputElement>) => this.setState(
    { name: e.currentTarget.value },
    this.focusTaskName,
  );

  /**
   * Edit the tag.
   *
   * @param {number} tag the new tag.
   */
  editTag = (tag: number) => this.setState({ tag, tagPickerOpened: false }, this.focusTaskName);

  /**
   * Edit the date.
   *
   * @param {Date} date the new date.
   */
  editDate = (date: Date) => this.setState(
    { date, datePickerOpened: false, datePicked: true },
    this.focusTaskName,
  );

  /**
   * Toggle the pin status.
   * @param {boolean} inFocus the new in-focus status.
   */
  togglePin = (inFocus: boolean) => this.setState({ inFocus }, this.focusTaskName);

  /**
   * Add a new subtask.
   *
   * @param e the event that contains the new name for new sub-task.
   */
  addNewSubTask = (e: SyntheticEvent<HTMLInputElement>) => {
    const newSubTaskName = e.currentTarget.value;
    if (newSubTaskName === '') {
      return;
    }
    this.setState(({ subtasks }: State) => ({
      subtasks: [...subtasks, {
        id: subtasks.length,
        name: newSubTaskName,
        complete: false,
        inFocus: false,
      }],
      needToSwitchFocus: true,
    }));
  };

  /**
   * Edit a subtask.
   *
   * @param {number} subtaskId id of the subtask to edit.
   * @return {Function<SyntheticEvent<HTMLInputElement>, void>} the event handler.
   */
  editSubTask = (subtaskId: number) => (e: SyntheticEvent<HTMLInputElement>) => {
    const name = e.currentTarget.value;
    this.setState(({ subtasks }: State) => ({
      subtasks: replaceSubTask(subtasks, subtaskId, s => ({ ...s, name })),
    }));
  };

  /**
   * Potentially submit a subtask.
   *
   * @param {SyntheticKeyboardEvent<HTMLInputElement>} e the keyboard event.
   */
  submitSubTask = (e: SyntheticKeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      this.handleSave();
    }
  };

  /**
   * Delete a subtask.
   *
   * @param {number} subtaskId id of the subtask to delete.
   * @return {Function<SyntheticEvent<HTMLInputElement>, void>} the event handler.
   */
  deleteSubTask = (subtaskId: number) => (e: SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    this.setState(({ subtasks }: State) => ({
      subtasks: subtasks.filter(s => s.id !== subtaskId),
    }));
  };

  /**
   * Reset the task.
   */
  resetTask = () => this.setState({ ...initialState() }, this.focusTaskName);

  addTask: ?HTMLInputElement;

  /**
   * Renders the editor for all the other info except main task name.
   *
   * @return {Node} the rendered other info editor.
   */
  renderOtherInfoEditor(): Node {
    const { opened } = this.state;
    if (!opened) {
      return null;
    }
    const {
      tag, date, inFocus, subtasks,
      tagPickerOpened, datePickerOpened, datePicked, needToSwitchFocus,
    } = this.state;
    const existingSubTaskEditor = ({ id, name }: SubTask, i: number, arr: SubTask[]) => {
      const refHandler = (inputElementRef) => {
        if (i === arr.length - 1 && needToSwitchFocus && inputElementRef != null) {
          inputElementRef.focus();
          this.setState({ needToSwitchFocus: false });
        }
      };
      return (
        <li key={id}>
          <button type="button" tabIndex={-1} onClick={this.deleteSubTask(id)}>
            <Icon name="delete" />
          </button>
          <input
            type="text"
            ref={refHandler}
            value={name}
            onChange={this.editSubTask(id)}
            onKeyDown={this.submitSubTask}
          />
        </li>
      );
    };
    return (
      <div className={styles.NewTaskActive}>
        <FocusPicker pinned={inFocus} onPinChange={this.togglePin} />
        <TagPicker
          tag={tag}
          opened={tagPickerOpened}
          onTagChange={this.editTag}
          onPickerOpened={this.openTagPicker}
        />
        <DatePicker
          date={date}
          opened={datePickerOpened}
          datePicked={datePicked}
          onDateChange={this.editDate}
          onPickerOpened={this.openDatePicker}
        />
        <button tabIndex={-1} type="submit" className={styles.SubmitNewTask}>
          <Icon name="arrow alternate circle right outline" color="black" />
        </button>
        <div className={styles.NewTaskModal}>
          <ul>{subtasks.map(existingSubTaskEditor)}</ul>
          <Icon name="plus" />
          <input type="text" placeholder="Add a Subtask" value="" onChange={this.addNewSubTask} />
          <button type="button" className={styles.ResetButton} onClick={this.resetTask}>
            {'Clear'}
          </button>
        </div>
      </div>
    );
  }

  render(): Node {
    const { name, opened } = this.state;
    const toggleDisplayStyle = opened ? {} : { display: 'none' };
    // Click this component, new task component closes.
    const newTaskCloser = (
      <div
        onClick={this.closeNewTask}
        role="presentation"
        className={styles.CloseNewTask}
        style={toggleDisplayStyle}
      />
    );
    const mainTaskNameEditor = (
      <input
        required
        type="text"
        value={name}
        onChange={this.editTaskName}
        className={styles.NewTaskComponent}
        placeholder={opened ? '' : PLACEHOLDER_TEXT}
        ref={(e) => { this.addTask = e; }}
      />
    );
    return (
      <div>
        {newTaskCloser}
        <form className={styles.NewTaskWrap} onSubmit={this.handleSave} onFocus={this.openNewTask}>
          {mainTaskNameEditor}
          {this.renderOtherInfoEditor()}
        </form>
      </div>
    );
  }
}

const actionCreators = { addTask: addTaskAction, removeTask: removeTaskAction };
const ConnectedTaskCreator = dispatchConnect<Props, typeof actionCreators>(
  actionCreators,
)(TaskCreator);
export default ConnectedTaskCreator;
