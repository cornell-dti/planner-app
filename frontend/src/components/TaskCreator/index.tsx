import React, { CSSProperties, KeyboardEvent, SyntheticEvent, ReactElement } from 'react';
import { connect } from 'react-redux';
import { randomId } from 'common/util/general-util';
import {
  Task,
  RepeatingDate,
  SamwiseUserProfile,
  SubTask,
  State as StoreState,
  Theme,
} from 'common/types/store-types';
import { NONE_TAG_ID } from 'common/util/tag-util';
import { isToday } from 'common/util/datetime-util';
import { subTasksEqual } from 'common/util/task-util';
import TagPicker from './TagPicker';
import DatePicker from './DatePicker';
import FocusPicker from './FocusPicker';
import GroupMemberPicker from './GroupMemberPicker';
import { addTask, TaskWithoutIdOrderChildren } from '../../firebase/actions';
import SamwiseIcon from '../UI/SamwiseIcon';
import styles from './index.module.scss';

type SimpleTask = Omit<Task, 'type' | 'order' | 'children' | 'metadata'>;

type State = SimpleTask & {
  readonly owner: readonly string[];
  readonly member?: SamwiseUserProfile;
  readonly date: Date | RepeatingDate;
  readonly subTasks: SubTask[];
  readonly opened: boolean;
  readonly tagPickerOpened: boolean;
  readonly datePickerOpened: boolean;
  readonly datePicked: boolean;
  readonly needToSwitchFocus: boolean;
};

type OwnProps = {
  readonly theme: Theme;
};

type Props = OwnProps & {
  readonly view: string;
  readonly group?: string;
  readonly groupMemberProfiles?: SamwiseUserProfile[];
  readonly taskCreatorOpened?: boolean;
  readonly assignedMember?: SamwiseUserProfile;
  readonly clearAssignedMember?: () => void;
};

/**
 * The placeholder text in the main task input box.
 */
const PLACEHOLDER_TEXT = 'What do you have to do?';
/**
 * Generate the initial state.
 */
const initialState = (): State => ({
  id: randomId(),
  owner: [''],
  name: '',
  tag: NONE_TAG_ID, // the id of the None tag.
  member: undefined,
  date: new Date(),
  complete: false,
  inFocus: false,
  subTasks: [],
  opened: false,
  tagPickerOpened: false,
  datePickerOpened: false,
  datePicked: false,
  needToSwitchFocus: false,
});

export class TaskCreator extends React.PureComponent<Props, State> {
  public readonly state: State = initialState();

  private addTask: HTMLInputElement | null | undefined;

  private get isOpen(): boolean {
    // eslint-disable-next-line react/destructuring-assignment
    return this.state.opened || this.props.taskCreatorOpened || false;
  }

  private get darkModeStyle(): CSSProperties | undefined {
    const { theme } = this.props;
    return theme === 'dark' ? { background: 'black', color: 'white' } : undefined;
  }

  /*
   * --------------------------------------------------------------------------------
   * Part 1: Openers & Closers
   * --------------------------------------------------------------------------------
   */

  private openNewTask = (): void => this.setState({ opened: true });

  private closeNewTask = (): void => this.setState({ opened: false });

  /**
   * Open the tag picker and close the date picker.
   */
  private openTagPicker = (): void =>
    this.setState(({ tagPickerOpened }: State) => ({
      tagPickerOpened: !tagPickerOpened,
      datePickerOpened: false,
    }));

  /**
   * Open the date picker and close the tag picker.
   */
  private openDatePicker = (): void =>
    this.setState(({ datePickerOpened }: State) => ({
      datePickerOpened: !datePickerOpened,
      tagPickerOpened: false,
    }));

  /*
   * --------------------------------------------------------------------------------
   * Part 2: Manager functions when finished editing.
   * --------------------------------------------------------------------------------
   */

  private focusTaskName = (): void => {
    if (this.addTask) {
      this.addTask.focus();
    }
  };

  private handleSave = (e?: SyntheticEvent<HTMLElement>): void => {
    if (e != null) {
      e.preventDefault();
    }
    const { owner, name, tag, date, complete, inFocus, subTasks } = this.state;
    if (name === '') {
      return;
    }
    const newSubTasks = subTasks.filter((subTask) => subTask.name !== ''); // remove empty subtasks
    // Put task in focus is the due date is today.
    const autoInFocus = inFocus || (date instanceof Date && isToday(date));
    const commonTask = { owner, name, tag, date, complete, inFocus: autoInFocus };
    let newTask: TaskWithoutIdOrderChildren;
    if (date instanceof Date) {
      date.setHours(23);
      date.setMinutes(59);
      date.setSeconds(59);
      const { view, group } = this.props;
      if (view === 'group' && typeof group === 'string') {
        newTask = { ...commonTask, metadata: { type: 'GROUP', date, group } };
      } else {
        newTask = { ...commonTask, metadata: { type: 'ONE_TIME', date } };
      }
    } else {
      newTask = {
        ...commonTask,
        inFocus: false,
        metadata: {
          type: 'MASTER_TEMPLATE',
          forks: [],
          date,
        },
      };
    }
    // Add the task to the store.
    addTask(owner, newTask, newSubTasks);
    // Reset the state.
    this.setState({ ...initialState() });
    this.closeNewTask();
    if (this.addTask) {
      this.addTask.blur();
    }
  };

  /*
   * --------------------------------------------------------------------------------
   * Part 3: Various Editors
   * --------------------------------------------------------------------------------
   */

  private editTaskName = (e: SyntheticEvent<HTMLInputElement>): void =>
    this.setState({ name: e.currentTarget.value }, this.focusTaskName);

  private editTag = (tag: string): void =>
    this.setState({ tag, tagPickerOpened: false }, this.focusTaskName);

  private editMember = (member?: SamwiseUserProfile): void =>
    this.setState({ member, tagPickerOpened: false }, this.focusTaskName);

  /**
   * Edit the date.
   *
   * @param {Date} date the new date, or null for cancel.
   */
  private editDate = (date: Date | RepeatingDate | null): void => {
    const { datePicked } = this.state;
    if (datePicked && date === null) {
      // User cancelled, but date was already picked
      this.setState({ datePickerOpened: false }, this.focusTaskName);
    } else if (date instanceof Date || date === null) {
      // Selecting a date, or user cancelled while date was not picked
      this.setState(
        {
          date: date ?? new Date(),
          datePickerOpened: false,
          datePicked: Boolean(date),
        },
        this.focusTaskName
      );
    } else {
      // Repeating task
      this.setState(
        {
          date,
          datePickerOpened: false,
          datePicked: true,
        },
        this.focusTaskName
      );
    }
  };

  private clearDate = (): void => {
    this.setState(
      { date: new Date(), datePickerOpened: false, datePicked: false },
      this.focusTaskName
    );
  };

  private togglePin = (inFocus: boolean): void => this.setState({ inFocus }, this.focusTaskName);

  private addNewSubTask = (e: SyntheticEvent<HTMLInputElement>): void => {
    const newSubTaskName = e.currentTarget.value;
    if (newSubTaskName === '') {
      return;
    }
    this.setState(({ subTasks }: State) => ({
      subTasks: [
        ...subTasks,
        {
          order: subTasks.reduce((acc, s) => Math.max(acc, s.order), 0) + 1,
          name: newSubTaskName,
          complete: false,
          inFocus: false,
        },
      ],
      needToSwitchFocus: true,
    }));
  };

  private newSubTaskKeyPress = (e: KeyboardEvent<HTMLInputElement>): void => {
    if (e.key === 'Tab') {
      this.closeNewTask();
    }
  };

  private editSubTask = (subTask: SubTask) => (e: SyntheticEvent<HTMLInputElement>) => {
    const name = e.currentTarget.value;
    this.setState(({ subTasks }: State) => ({
      subTasks: subTasks.map((s) => (subTasksEqual(s, subTask) ? { ...s, name } : s)),
    }));
  };

  private submitSubTask = (e: KeyboardEvent<HTMLInputElement>): void => {
    if (e.key === 'Enter') {
      this.handleSave();
    }
  };

  private deleteSubTask = (subTask: SubTask) => (e: SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    this.setState(({ subTasks }: State) => ({
      subTasks: subTasks.filter((s) => !subTasksEqual(s, subTask)),
    }));
  };

  private resetTask = (): void => this.setState({ ...initialState() }, this.focusTaskName);

  public render(): ReactElement {
    const { view, groupMemberProfiles, assignedMember, clearAssignedMember } = this.props;
    const {
      name,
      tag,
      member,
      date,
      inFocus,
      subTasks,
      tagPickerOpened,
      datePickerOpened,
      datePicked,
      needToSwitchFocus,
    } = this.state;
    if (!this.isOpen) {
      return (
        <div
          className={`${styles.TaskCreator} ${styles.TaskCreatorClosed}`}
          style={this.darkModeStyle}
        >
          <form
            className={`${styles.NewTaskWrap} ${groupMemberProfiles ? styles.GroupTaskWrap : ''}`}
            onSubmit={this.handleSave}
            onFocus={this.openNewTask}
          >
            <input
              required
              type="text"
              value={name}
              onChange={this.editTaskName}
              className={styles.NewTaskComponent}
              placeholder={PLACEHOLDER_TEXT}
              style={this.darkModeStyle}
            />
          </form>
        </div>
      );
    }

    const existingSubTaskEditor = (
      thisSubTask: SubTask,
      i: number,
      arr: SubTask[]
    ): ReactElement => {
      const refHandler = (inputElementRef: HTMLInputElement | null): void => {
        if (i === arr.length - 1 && needToSwitchFocus && inputElementRef != null) {
          inputElementRef.focus();
          this.setState({ needToSwitchFocus: false });
        }
      };

      const { order, name: subtaskName } = thisSubTask;
      return (
        <li key={order} className={styles.ExistingSubTaskRow}>
          <button
            className={styles.DeleteSubTaskButton}
            type="button"
            tabIndex={-1}
            onClick={this.deleteSubTask(thisSubTask)}
          >
            <SamwiseIcon iconName="x-dark" />
          </button>
          <input
            type="text"
            className={styles.ExistingSubTaskInput}
            ref={refHandler}
            value={subtaskName}
            onChange={this.editSubTask(thisSubTask)}
            onKeyDown={this.submitSubTask}
            style={this.darkModeStyle}
          />
        </li>
      );
    };

    return (
      <div className={styles.TaskCreator} style={this.darkModeStyle}>
        <div onClick={this.closeNewTask} role="presentation" className={styles.CloseNewTask} />
        <div className={styles.TaskCreatorOpenedPlaceHolder} />
        <form
          className={`${styles.NewTaskWrap} ${styles.NewTaskModal}`}
          onSubmit={this.handleSave}
          onFocus={this.openNewTask}
        >
          <div className={`${styles.TaskCreatorRow} ${styles.FirstRow}`}>
            <div className={styles.TitleText}>Add Task</div>
            {date instanceof Date && <FocusPicker pinned={inFocus} onPinChange={this.togglePin} />}
            <div className={styles.TagPickWrap}>
              {view === 'personal' ? (
                <TagPicker
                  tag={tag}
                  opened={tagPickerOpened}
                  onTagChange={this.editTag}
                  onPickerOpened={this.openTagPicker}
                />
              ) : (
                <GroupMemberPicker
                  member={assignedMember || member}
                  opened={tagPickerOpened}
                  onMemberChange={this.editMember}
                  onPickerOpened={this.openTagPicker}
                  groupMemberProfiles={groupMemberProfiles || []}
                  clearAssignedMember={clearAssignedMember}
                />
              )}
            </div>
            <DatePicker
              date={date}
              opened={datePickerOpened}
              datePicked={datePicked}
              inGroupView={false}
              onDateChange={this.editDate}
              onClearPicker={this.clearDate}
              onPickerOpened={this.openDatePicker}
            />
          </div>
          <div className={styles.TaskCreatorRow}>
            <input
              required
              type="text"
              value={name}
              onChange={this.editTaskName}
              className={`${styles.NewTaskComponent} ${styles.NewTaskComponentOpened}`}
              ref={(e) => {
                this.addTask = e;
              }}
              style={this.darkModeStyle}
            />
            <button
              type="submit"
              className={view === 'personal' ? styles.SubmitNewTask : styles.GroupSubmitNewTask}
              style={this.darkModeStyle}
            >
              <SamwiseIcon iconName="add-task" />
            </button>
          </div>
          <div className={styles.SubtitleText}>
            <b>Add Subtasks</b>
            &nbsp;(optional)
          </div>
          <div className={styles.DescText}>
            Add optional subtasks to break down your tasks into more manageable pieces.
          </div>
          <div style={this.darkModeStyle}>
            <div className={styles.SubtasksContainer}>
              <ul className={styles.SubtasksList}>{subTasks.map(existingSubTaskEditor)}</ul>
              <SamwiseIcon iconName="edit" containerClassName={styles.EditIcon} tabIndex={-1} />
              <input
                className={styles.SubtaskInput}
                type="text"
                placeholder="Add a Subtask"
                value=""
                onChange={this.addNewSubTask}
                onKeyDown={this.newSubTaskKeyPress}
                style={this.darkModeStyle}
              />
            </div>
          </div>
          <button type="button" className={styles.ResetButton} onClick={this.resetTask}>
            DISCARD TASK
          </button>
        </form>
      </div>
    );
  }
}

const Connected = connect(({ settings: { theme } }: StoreState): OwnProps => ({ theme }))(
  TaskCreator
);
export default Connected;
