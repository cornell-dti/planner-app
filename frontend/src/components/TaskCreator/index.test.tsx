import React from 'react';
import renderer from 'react-test-renderer';
import TaskCreator from '.';

it('TaskCreator matches snapshot', () => {
  const tree = renderer.create(<TaskCreator />).toJSON();
  expect(tree).toMatchSnapshot();
});
