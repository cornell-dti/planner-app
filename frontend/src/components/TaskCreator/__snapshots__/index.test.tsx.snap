// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCreator matches snapshot 1`] = `
<div
  className="TaskCreator"
>
  <div
    className="CloseNewTask"
    onClick={[Function]}
    role="presentation"
    style={
      Object {
        "display": "none",
      }
    }
  />
  <form
    className="NewTaskWrap"
    onFocus={[Function]}
    onSubmit={[Function]}
  >
    <input
      className="NewTaskComponent"
      onChange={[Function]}
      placeholder="What do you have to do?"
      required={true}
      type="text"
      value=""
    />
  </form>
</div>
`;
